utils/frame/dry-run-runtime-upgrade/cli/src/lib.rs

```diff
@@ -0,0 +1,140 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::{fmt::Debug, str::FromStr};
+use sc_service::Configuration;
+use sc_cli::{CliConfiguration, ExecutionStrategy};
+use sc_executor::{WasmExecutionMethod, NativeExecutor};
+use sp_state_machine::StateMachine;
+use sc_service::NativeExecutionDispatch;
+use sp_runtime::traits::{Block as BlockT, NumberFor};
+use sp_core::storage::{StorageData, StorageKey, well_known_keys};
+
+#[derive(Debug, structopt::StructOpt)]
+pub struct DryRunCmd {
+	/// The shared parameters
+	#[allow(missing_docs)]
+	#[structopt(flatten)]
+	pub shared_params: sc_cli::SharedParams,
+
+	/// The target pallet to run the migration against.
+	#[structopt(short, long, default_value = "All")]
+	pub target: Target,
+
+	/// The state to use to run the migration. It should be a ":" separated string, where the
+	/// prefix is either `live` or `snap`, and the postfix is either the HTTP uri or the file path,
+	/// respectively.
+	#[structopt(short, long, default_value = "live:http://localhost:9933")]
+	pub state: State,
+}
+
+/// Possible targets for dryrun runtime upgrade.
+#[derive(Debug)]
+pub enum Target {
+	/// All pallets.
+	All,
+	/// A single pallet.
+	Pallet(String),
+}
+
+impl FromStr for Target {
+	type Err = &'static str;
+	fn from_str(s: &str) -> Result<Self, Self::Err> {
+		Ok(if s.to_lowercase() == "All" { Target::All } else { Target::Pallet(s.to_string()) })
+	}
+}
+
+/// The state to use for a migration dryrun.
+#[derive(Debug)]
+pub enum State {
+	/// A snapshot. Inner value is file path.
+	Snap(String),
+
+	/// A live chain. Inner value is the HTTP uri.
+	Live(String),
+}
+
+impl FromStr for State {
+	type Err = &'static str;
+	fn from_str(s: &str) -> Result<Self, Self::Err> {
+		let splitted = s.splitn(2, ":").collect::<Vec<_>>();
+		if splitted.len() != 2 {
+			return Err("invalid format. Must be [state_type]:[state_value].");
+		}
+		let state_type = splitted[0];
+		let value = splitted[1];
+		match state_type {
+			"live" => Ok(State::Live(value.to_string())),
+			"snap" => Ok(State::Snap(value.to_string())),
+			_ => Err("Invalid state type."),
+		}
+	}
+}
+
+impl DryRunCmd {
+	pub async fn run<B, ExecDispatch>(&self, config: Configuration) -> sc_cli::Result<()>
+	where
+		B: BlockT,
+		ExecDispatch: NativeExecutionDispatch + 'static,
+	{
+		let spec = config.chain_spec;
+		let genesis_storage = spec.build_storage()?;
+
+		let code = StorageData(genesis_storage.top.get(well_known_keys::CODE).unwrap().to_vec());
+		let code_key = StorageKey(well_known_keys::CODE.to_vec());
+
+		let wasm_method = WasmExecutionMethod::Interpreted;
+		let strategy = ExecutionStrategy::Native;
+		let mut changes = Default::default();
+
+		let heap_pages = Some(1024);
+		let executor = NativeExecutor::<ExecDispatch>::new(wasm_method, heap_pages, 2);
+
+		let ext = remote_externalities::Builder::new().inject(&[(code_key, code)]).build().await;
```

Comment:
this `.inject` here is where the whole crux of the work happens: inject the _real_ code (i.e. the one that is currently in _this_ repo) into the remote _fake_ state. I should double-check that this is what we want.